README.md
Summary
----------------
Welcome to isometric. To install this python isometric 2d engine, you will need python, and pygame installed. This project is cross platform compatible, for Windows 7 (tested), Linux, and Mac OS X (Yosemite tested).

For reading about good programming guidelines, see:
https://www.pygame.org/docs/tut/tom/MakeGames.html
http://pythonprogramming.net/pygame-start-menu-tutorial/

Installation
-----------------

Windows:
---------

Step 1:
You can download Python at the official site here:
https://www.python.org/downloads/windows/

Download and install version 2.7.x or 3.x.

Step 2:
Download pygame from here:
http://www.lfd.uci.edu/~gohlke/pythonlibs/#pygame

Once downloaded, open up a command prompt and install with pip, giving your custom path to the wh1 file below:
```ssh
py -m pip install C:/Path/to/pygame-1.9.2a0-cp34-none-win_amd64.whl
```

Step 3:
Run app.py with PyCharm or from the command line.


Mac:
----------
Step 1:
Install all Pygame deps for Mac with Brew (if you don't have brew, look up how to install first)

```ssh
    brew install python3 hg sdl sdl_image sdl_ttf portmidi smpeg 
```

Step 2:
Install PyGame
```ssh
    pip3 install hg+http://bitbucket.org/pygame/pygame
```

Step 3:
Download sdl-mixer source code from: https://www.libsdl.org/projects/SDL_mixer/release-1.2.html

Untar the tar.gz file with tar zxvf:
```ssh
    tar zxvf SDL_mixer-1.2.12.tar.gz
    cd SDL_mixer-1.2.12
    ./configure
    make
    sudo make install
```

Step 4:
Run app.py and see if everything works!
```ssh
python app.py
```


Ubuntu (14):
----------
Step 1:
Install all Pygame deps for Linux with yum / apt-get

```ssh
    sudo apt-get install python3 mercurial python3-pip libfreetype6-dev 
    sudo apt-get build-dep ptyhon-pygame
```

Step 2:
Install PyGame
```ssh
    sudo pip3 install hg+http://bitbucket.org/pygame/pygame
```

Step 3:
Download sdl-mixer source code from: https://www.libsdl.org/projects/SDL_mixer/release-1.2.html

Untar the tar.gz file with tar zxvf:
```ssh
    tar zxvf SDL_mixer-1.2.12.tar.gz
    cd SDL_mixer-1.2.12
    ./configure
    make
    sudo make install
```

Step 4:
Run app.py and see if everything works!
```ssh
python app.py
```

RoadMap
---------

~~Random Map~~

~~Health Bar~~

Load/Save Game

Inventory Menu

Inventory

Animated Character Sprite (~~walking~~, standing, attack)

Animated Enemy Sprite

Inventory Hot Bar

Tile placement function

Light Shadows

Decent BG Music

SFX Package

Draw Distance Detection and Rendering


changelog
0.1.2 - Fixed bugs with walking off of the edge of the map
      - Fixed positioning so that collisions make sense
      - Fixed regen to work better on 8 bit/fps
      - Updated to 16 bit, as it looks better


0.1.1 - Added a Main menu
      - Reenabled character walk sounds
      - Setup info toggle (i for now, until inventory is implemented)
      - Setup player sprite and itemSprite objects to manage the game's sprites.
      - Setup probability for maps to be able to adjust difficulty in different areas

0.1.0 - Added a player sprite with no animation, with a left and right orientation.
      - Can move character
      - Map will move when you use WASD or the arrow keys
      - Added an animated sprite to the map
      - Added 4 main tiles: grass, water, stone, and wood floor
      - Setup configuration map sizes
      - Created a random map function

0.1 - Started the project. Start with tile layouts from a stackoverflow question


unit_tests.py
# Let's write some tests for a nightly build

import unittest
from app import gameloop

class TestGameLoop(unittest.TestCase):
    def test_gameloop(self):
      #self.assertTrue(gameloop())
      with self.assertRaises(SystemExit):
        gameloop()

menus.py
import pygame, sys
from pygame.locals import *
from config import display_height, display_width, version, game_font

#from ui import button

pygame.init()
display = pygame.display.set_mode((display_height,display_width))
pygame.display.set_caption('Isometric v%s' % version)
start = [pygame.image.load('images/obj/button.png').convert(), pygame.image.load('images/obj/button_i.png').convert()]
menubg = pygame.image.load('images/obj/mainmenubg.png').convert()


# Assets
menu_music = pygame.mixer.music.load('audio/sfx/menu_screen.mp3')
menu_item = pygame.mixer.Sound('audio/sfx/menu_item.ogg')



def quit():
  pygame.quit()
  sys.exit()

def esc_menu(display):
  #escSurf = game_font.rander('')
  menu_bg = pygame.Rect(300,200,200,200)
  pygram.draw.rect(display, (0,0,0), menu_bg)
  button("Back",300,300,200,50,start[0],start[1],self.close_menu()),


  def close_menu():
      # Clean up?? we dont really need to do anything unless it frees ram
      b=22



def mainmenu(main):
  titleFont = pygame.font.Font('fonts/04B_30.ttf',72)
  #subFont =
  titleSurf = titleFont.render('Isometric', True, (0,0,0))
  #subSurf = subFont.render('by Ryan Hartje', True, (0,0,0))
  titleRect = titleSurf.get_rect()
  #subRect = subSurf.get_rect()
  titleRect.center = ((400,100))
  #subRect.center = ((400,200))

  # MUTED, save me
  #pygame.mixer.music.play(0)
  while True:
    for event in pygame.event.get():
      # This is for when you click X on the game window
      if event.type == pygame.QUIT:
        quit()

      # And this is for when you hit Escape in the Menu screen.
      keys = pygame.key.get_pressed()
      if keys[K_ESCAPE]:
        esc_menu()

    # menu
    display.fill((225,225,225))

    # draw repeating isometric background
    for y in range(0,display_width, 180):
      for x in range(0, display_height, 104):
        display.blit(menubg, (x,y))


    display.blit(titleSurf, titleRect)
    #display.blit(subSurf, subRect)


    button("Start",300,300,200,50,start[0],start[1],main)
    pygame.display.update()

# label, x,y, width, height, inactive_color, active_color, action or subprogram
def button(msg,x,y,w,h,inactive_image, active_image,action=None):
    played=False
    mouse = pygame.mouse.get_pos()
    click = pygame.mouse.get_pressed()
    if x+w > mouse[0] > x and y+h > mouse[1] > y:
        if not played:
          #buggy
          #menu_item.play()
          played=True
        #pygame.draw.rect(display, ac,(x,y,w,h))
        display.blit(active_image, (x,y))

        if click[0] == 1 and action != None:
            action()
    else:
        #pygame.draw.rect(display, inactive_image,(x,y,w,h))
        display.blit(active_image, (x,y))
        played=False
    smallText = pygame.font.Font("fonts/Minecraft.ttf",26)

    textSurf = smallText.render(msg, True, (245,245,245))
    textRect = textSurf.get_rect()
    textRect.center = ( (x+(w/2)), (y+(h/2)) )
    display.blit(textSurf, textRect)


assets.py

map = [
[0,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,2,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,2,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
]





mapDraw.py
# from assets import map
#
import random

size = 64

def randomMap(probability, list_obj):
  # random map
  rmap = []
  for x in range(0,size):
    rmap.append([])
    for y in range(0,size):

      # Let's make sure at least 85% of our tiles are walkable
      dice = random.randint(0,100)
      if dice>probability:
        # in the 15% of the time we do gen a random tile,
        #  let's try to clump similar types together
        if not y > 0 or random.randint(1,10) < 5:
          random_item = int(random.randrange(0, len(list_obj)))
        else:
          if random.randint(0,1) == 0 or x == 0:
            if rmap[x][y-1] == 1 or rmap[x][y-1] == 2:
              random_item = rmap[x][y-1]
            else:
              random_item = 2
          else:
            random_item = rmap[x-1][y]
        rmap[x].append(random_item)
      else:
        rmap[x].append(0)
  return rmap

def drawMap(xoffset,yoffset,tiles,display,map,items):

  currentRow = 0 # y
  currentTile = 0 # x

  # Begin drawing the map, starting at [0,0] in the map data set
  for row in map:
    for tile in row:
      tileImage = tiles[map[currentRow][currentTile]]
      tileImage.set_colorkey((0,0,0))
  
      cartx = currentTile*64
      carty = currentRow*64

      x = xoffset + 0 + ((cartx - carty) / 2)
      y = yoffset + 0 + ((cartx+carty) / 4)
  
      currentTile+=1

      # draw to display
      display.blit(tileImage, (x,y))
      #coins = [
      #  itemSprite((500,300)),
      #]


    
    currentTile = 0 # Reset for new row
    currentRow += 1
 


app.py
import pygame, sys, random
from pygame.locals import *

from config import display_height, display_width, version, save, fps_setting
from menus import mainmenu, esc_menu
from mapDraw import drawMap, randomMap




class playerSprite(pygame.sprite.Sprite):
  frames = [ pygame.image.load('images/obj/player/character_stand_l.png'),
             pygame.image.load('images/obj/player/character_stand_l2.png'),
             pygame.image.load('images/obj/player/character_stand_r.png'),
             pygame.image.load('images/obj/player/character_stand_r2.png')]

  def __init__(self):
    pygame.sprite.Sprite.__init__(self)
    self.rect = pygame.Rect(self.frames[0].get_rect())
    self.rect.center = (420,372)
    self.anim = 0
    self.image = self.frames[0]

  def update(self, direction):
    if direction == 'u' or direction == 'l':
      self.anim = random.randrange(0,2)
    else:
      self.anim = random.randrange(2,4)
    self.image = self.frames[self.anim]


class itemSprite(pygame.sprite.Sprite):

  def __init__(self, position,frames):
    pygame.sprite.Sprite.__init__(self)
    self.frames = frames
    self.rect = pygame.Rect(self.frames[0].get_rect())
    self.rect.center = position
    self.anim = 0

  def update(self, hitlist):
    if self in hitlist: self.image = None
    else: 
      self.anim += 1
      if self.anim >= len(self.frames): self.anim = 0
      self.image = self.frames[self.anim]


def gameloop():
  # Load Resources
  pygame.init()
  display = pygame.display.set_mode((display_height,display_width), DOUBLEBUF)
  pygame.display.set_caption("Isometric")
  clock = pygame.time.Clock()
  pygame.mixer.init()
  #mapmusic = pygame.mixer.music.load('audio/music/.mp3')
  #mapmusic = pygame.mixer.music.load('audio/sfx/menu_screen.mp3')
  walk = [pygame.mixer.Sound('audio/sfx/footstep01.ogg'),pygame.mixer.Sound('audio/sfx/footstep01.ogg')]
  font = pygame.font.Font('fonts/Minecraft.ttf', 16)
  
  # Load all frames of animation for the player
  player = [ pygame.image.load('images/obj/player/character_stand_l.png').convert(),
             pygame.image.load('images/obj/player/character_stand_l2.png').convert(),
             pygame.image.load('images/obj/player/character_stand_r.png').convert(),
             pygame.image.load('images/obj/player/character_stand_r2.png').convert()]

  tiles = [pygame.image.load('images/tiles/grass.png').convert(), pygame.image.load('images/tiles/wall.png').convert(),
           pygame.image.load('images/tiles/water.png').convert(), pygame.image.load('images/tiles/wood.png').convert()]

  coin_imgs =  [ pygame.image.load('images/obj/coin/coin_1.png'),
    pygame.image.load('images/obj/coin/coin_2.png'),
    pygame.image.load('images/obj/coin/coin_3.png'),
    pygame.image.load('images/obj/coin/coin_4.png') ]

  blocking_tiles = [1,2]
  bg = pygame.image.load('images/bgs/stardust.png').convert()

  
  
  # Generate our map
  map = randomMap(85,tiles)

  # Setup our sprites
  players = [ playerSprite(), ]
  sprites = [ itemSprite((0,0),coin_imgs)]

  # This should be moved to the map drawing to handle coordinates for x,y and offset.
  #coins = [
  #  itemSprite((500,300)),
  #]

 
  # Render our groups to the display
  player_group = pygame.sprite.RenderPlain(*players)
  coin_group = pygame.sprite.RenderPlain(*sprites)
  
  if not save:
    # Setup Variables
    xoffset = 0
    yoffset = 0
    xpos = 18
    ypos = 6
    
    health=60
    stamina=40
    regen=1
    
    inventory = {}

    info_toggle = False
    esc_toggle  = False
  
  # Event loop
  # pygame.mixer.music.play(0)
  while True:



    for event in pygame.event.get():
      # Mouse Bindings
      if event.type == MOUSEBUTTONDOWN:
        #sound.play()
        mx, my = pygame.mouse.get_pos()
        print("%s,%s" % (mx, my))
      if event.type == KEYDOWN:
          if event.key == K_i:
            if info_toggle:
              info_toggle = False
            else:
              info_toggle = True
          if event.key == K_ESCAPE:
            pygame.quit()
            sys.exit(1)





    # Key Bindings
    keys = pygame.key.get_pressed()


    if not ypos == len(map)-1 and keys[K_LEFT] or keys[K_a] and keys[K_LSHIFT]:
      # BUGGY if you walk off of the map
      if map[ypos+1][xpos] not in blocking_tiles:
        xoffset += 32
        yoffset -= 16
        ypos += 1
        walk[random.randrange(2)].play()
        player_group.update('l')

    if ypos>0 and keys[K_RIGHT] or keys[K_d]:
      if map[ypos-1][xpos] not in blocking_tiles:
        yoffset += 16
        xoffset -= 32
        ypos -= 1
        walk[random.randrange(2)].play()
        player_group.update('r')

    if xpos>0 and keys[K_UP] or keys[K_w]and keys[K_LSHIFT]:
      if map[ypos][xpos-1] not in blocking_tiles:
        yoffset += 16
        xoffset += 32
        xpos -= 1
        walk[random.randrange(2)].play()
        player_group.update('u')

    if not xpos == len(map[0])-1 and keys[K_DOWN] or keys[K_s] and keys[K_LSHIFT]:
      if map[ypos][xpos+1] not in blocking_tiles:
        yoffset -= 16
        xoffset -= 32
        xpos += 1
        walk[random.randrange(2)].play()
        player_group.update('d')


  
  
    if health<100:
      health += regen

    if stamina<100:
      stamina += regen
  
    # Setup frame
    display.fill((0,0,0)) # fill to clear frame
    display.blit(bg,(0,0)) # background
    drawMap(xoffset, yoffset, tiles, display, map,sprites) # render the map
    #display.blit(player[0], (385,336)) # draw our character change to sprite later
  
    # health bar horizontal
    pygame.draw.rect(display, (0,0,0), pygame.Rect(22,20,106,12))
    pygame.draw.rect(display, (200,0,0), pygame.Rect(25,22,health,8))

    # Stamina
    pygame.draw.rect(display, (0,0,0), pygame.Rect(22,36,106,12))
    pygame.draw.rect(display, (200,200,200), pygame.Rect(25,38,stamina,8))
  
    # Draw sprites to screen
    collisions = pygame.sprite.spritecollide(players[0], coin_group, True)
    coin_group.update(collisions)
    #coin_group.draw(display)

    player_group.draw(display)

    # Escape menu
    if esc_toggle:
      esc_menu(display)


    # Info
    if info_toggle:

      # label and version
      label = font.render('isometric v%s' % version, True, (250,250,250))
      label_obj = label.get_rect()
      label_box = label_obj.center = (680,580)
      display.blit(label, label_box)

      info = font.render("(%s,%s): Tile %s" % (xpos,ypos,map[ypos][xpos]), True, (250,250,250))
      info_obj = info.get_rect()
      info_box = label_obj.center = (25,580)
      display.blit(info,info_box)
  
      fps = font.render("fps: %s" % (int(clock.get_fps())), True, (250,250,250))
      fps_obj = fps.get_rect()
      fps_box = label_obj.center = (725,25)
      display.blit(fps,fps_box)


  
    clock.tick(fps_setting)
    pygame.display.update()


if __name__ == "__main__":
  mainmenu(gameloop)
